src/eos-updater-poll-lan.c

```diff
@@ -0,0 +1,920 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright 2016 Kinvolk GmbH
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Author: Krzesimir Nowak <krzesimir@kinvolk.io>
+ */
+
+#include "eos-updater-avahi.h"
+#include "eos-updater-poll-common.h"
+#include "eos-updater-poll-lan.h"
+
+#include "eos-util.h"
+
+#include <string.h>
+#include <errno.h>
+
+typedef enum
+{
+  TXT_RECORD_OK,
+  TXT_RECORD_NOT_UNIQUE,
+  TXT_RECORD_NOT_FOUND
+} TxtRecordError;
+
+typedef struct LanData LanData;
+
+struct LanData
+{
+  EosMetadataFetchData *fetch_data;
+  GMainLoop *main_loop;
+  GError *error;
+  EosUpdateInfo *info;
+  EosMetricsInfo *metrics;
+};
+
+#define LAN_DATA_CLEARED { NULL, NULL, NULL, NULL, NULL }
+
+static void
+lan_data_init (LanData *lan_data,
+               EosMetadataFetchData *fetch_data)
+{
+  if (lan_data == NULL)
+    return;
+
+  g_return_if_fail (EOS_IS_METADATA_FETCH_DATA (fetch_data));
+
+  memset (lan_data, 0, sizeof (*lan_data));
+  lan_data->fetch_data = g_object_ref (fetch_data);
+  lan_data->main_loop = g_main_loop_new (fetch_data->context, FALSE);
+}
+
+static void
+lan_data_clear (LanData *lan_data)
+{
+  if (lan_data == NULL)
+    return;
+
+  g_clear_object (&lan_data->metrics);
+  g_clear_object (&lan_data->info);
+  g_clear_error (&lan_data->error);
+  g_clear_pointer (&lan_data->main_loop, g_main_loop_unref);
+  g_clear_object (&lan_data->fetch_data);
+}
+
+G_DEFINE_AUTO_CLEANUP_CLEAR_FUNC (LanData, lan_data_clear)
+
+static TxtRecordError
+fill_txt_records (gchar **txt_records,
+                  GHashTable *keys)
+{
+  gchar **iter;
+  g_autoptr(GPtrArray) keys_to_find = g_ptr_array_new ();
+  const gchar *fail_key = NULL;
+  TxtRecordError error = TXT_RECORD_OK;
+
+  {
+    GHashTableIter key_iter;
+    gpointer key;
+
+    g_hash_table_iter_init (&key_iter, keys);
+    while (g_hash_table_iter_next (&key_iter, &key, NULL))
+      g_ptr_array_add (keys_to_find, key);
+  }
```

Comment:
Only to limit the use of `key_iter` and `key` variables. I'll change that a bit.